File: matlab/ch08/ellipticPDE.m
function [x,y,U] = ellipticPDE(nx,ny,dx,dy,bc,f)
% Solve 2-dimensional elliptic PDE
% output:
% x, y: vectors of x and y values
% U: matrix of dependent variables (U(x,y))
% input:
% nx, ny: number of divisions in x and y direction
% dx, dy: x and y increments
% f: constant for Poisson equation
% bc: a matrix containing types and values of boundary conditions
% in x and y directions (4x2 or 4x3 matrix)
% order of appearing: lower x, upper x, lower y, upper y
% in rows 1 to 4 of the matric bc
% 1st column: type of condition:
% (1: Dirichlet condition, values of u in 2nd column
% 2: Neumann condition, values of u' in 2nd column
% 3: Robbins condition, 2-3 columns contain constant and coef. of u)

% Initialization
if nargin < 5, error('Invalid number of inputs.'); end
[a,b] = size(bc);
if a ~= 4, error('Invalid number of boundary conditions.'); end
if b < 2 | b > 3, error('Invalid boundary condition.'); end
if b == 2 & max(bc(:,1)) <= 2, bc = [bc zeros(4,1)]; end
if nargin < 6 | isempty(f), f = 0; end
nx = fix(nx); x = [0:nx]*dx; ny = fix(ny); y = [0:ny]*dy; dx2 = 1/dx^2; dy2 = 1/dy^2;
% Coefficient matrix and constant vector
n = (nx+1)*(ny+1); A = zeros(n); c = zeros(n,1);
onex = diag(diag(ones(nx-1))); oney = diag(diag(ones(ny-1)));
% Interior nodes
i = [2:nx];
for j = 2:ny
ind = (j-1)*(nx+1)+i; A(ind,ind) = -2*(dx2+dy2)*onex;
A(ind,ind+1) = A(ind,ind+1) + dx2*onex; A(ind,ind-1) = A(ind,ind-1) + dx2*onex;
A(ind,ind+nx+1) = A(ind,ind+nx+1) + dy2*onex;
A(ind,ind-nx-1) = A(ind,ind-nx-1) + dy2*onex; c(ind) = f*ones(nx-1,1);
end
% Lower x boundary condition
switch bc(1,1)
case 1
ind = ([2:ny]-1)*(nx+1)+1; A(ind,ind) = A(ind,ind) + oney; c(ind) = bc(1,2)*ones(ny-1,1);
case {2, 3}
ind = ([2:ny]-1)*(nx+1)+1; A(ind,ind) = A(ind,ind) - (3/(2*dx) + bc(1,3))*oney;
A(ind,ind+1) = A(ind,ind+1) + 2/dx*oney; A(ind,ind+2) = A(ind,ind+2) - 1/ (2*dx)*oney;
c(ind) = bc(1,2)*ones(ny-1,1);
end
% Upper x boundary condition
switch bc(2,1)
case 1
ind = [2:ny]*(nx+1); A(ind,ind) = A(ind,ind) + oney; c(ind) = bc(2,2)*ones(ny-1,1);
case {2, 3}
ind = [2:ny]*(nx+1); A(ind,ind) = A(ind,ind) + (3/(2*dx) - bc(2,3))*oney;
A(ind,ind-1) = A(ind,ind-1) - 2/dx*oney;
A(ind,ind-2) = A(ind,ind-2) + 1/(2*dx)*oney; c(ind) = bc(2,2)*ones(ny-1,1);
end
% Lower y boundary condition
switch bc(3,1)
case 1
ind = [2:nx]; A(ind,ind) = A(ind,ind) + onex; c(ind) = bc(3,2)*ones(nx-1,1);
case {2, 3}
ind = [2:nx]; A(ind,ind) = A(ind,ind) - (3/(2*dy) + bc(3,3))*onex;
A(ind,ind+nx+1) = 2/dy*onex; A(ind,ind+2*(nx+1)) = -1/(2*dy)*onex;
c(ind) = bc(3,2)*ones(nx-1,1);
end
% Upper y boundary condition
switch bc(4,1)
case 1
ind = ny*(nx+1)+[2:nx]; A(ind,ind) = A(ind,ind) + onex; c(ind) = bc(4,2)*ones(nx-1,1);
case {2, 3}
ind = ny*(nx+1)+[2:nx]; A(ind,ind) = A(ind,ind) + (3/(2*dy) - bc(4,3))*onex;
A(ind,ind-(nx+1)) = A(ind,ind-(nx+1)) - 2/dy*onex;
A(ind,ind-2*(nx+1)) = A(ind,ind-2*(nx+1)) + 1/(2*dy)*onex; c(ind) = bc(4,2)*ones(nx-1,1);
end
% Corner nodes
A(1,1) = 1; A(1,2) = -1/2; A(1,nx+2) = -1/2; c(1) = 0;
A(nx+1,nx+1) = 1; A(nx+1,nx) = -1/2; A(nx+1,2*(nx+1)) = -1/2; c(nx+1) = 0;
A(ny*(nx+1)+1,ny*(nx+1)+1) = 1; A(ny*(nx+1)+1,ny*(nx+1)+2) = -1/2;
A(ny*(nx+1)+1,(ny-1)*(nx+1)+1) = -1/2; c(ny*(nx+1)+1) = 0;
A(n,n) = 1; A(n,n-1) = -1/2; A(n,n-(nx+1)) = -1/2; c(n) = 0;
u = inv(A)*c; % solve equation
% Rearrange results in matrix form
for k = 1:ny+1, U(k,1:nx+1) = u((k-1)*(nx+1)+1:k*(nx+1))'; end
end

File: python/ch08/ellipticPDE.py
import numpy as np

def ellipticPDE(nx, ny, dx, dy, bc, f=0):
    """
    Solve 2-dimensional elliptic PDE (Python version of ellipticPDE.m)
    
    Args:
        nx, ny: number of divisions in x and y direction
        dx, dy: x and y increments
        bc: matrix of boundary conditions (4x2 or 4x3)
        f: constant for Poisson equation (default is 0 for Laplace)
    """
    
    # Initialization and validation
    bc = np.array(bc)
    a, b = bc.shape
    if a != 4:
        raise ValueError('Invalid number of boundary conditions.')
    if b < 2 or b > 3:
        raise ValueError('Invalid boundary condition.')
        
    # If bc is 4x2 and only contains Dirichlet/Neumann (types 1 or 2), pad with zeros
    if b == 2:
        bc = np.hstack((bc, np.zeros((4, 1))))
    
    nx = int(nx)
    ny = int(ny)
    x = np.arange(nx + 1) * dx
    y = np.arange(ny + 1) * dy
    dx2 = 1 / (dx**2)
    dy2 = 1 / (dy**2)
    
    # Coefficient matrix and constant vector
    n = (nx + 1) * (ny + 1)
    A = np.zeros((n, n))
    c = np.zeros(n)
    
    # Interior nodes
    # MATLAB indices 2:nx correspond to Python 1:nx
    for j in range(1, ny):
        for i in range(1, nx):
            ind = j * (nx + 1) + i
            A[ind, ind] = -2 * (dx2 + dy2)
            A[ind, ind + 1] = dx2
            A[ind, ind - 1] = dx2
            A[ind, ind + (nx + 1)] = dy2
            A[ind, ind - (nx + 1)] = dy2
            c[ind] = f

    # Boundary conditions - Helper for indexing
    # Matlab 1-based indexing to Python 0-based conversion
    
    # 1. Lower x boundary (i=0)
    for j in range(1, ny):
        ind = j * (nx + 1) + 0
        if bc[0, 0] == 1: # Dirichlet
            A[ind, ind] = 1.0
            c[ind] = bc[0, 1]
        elif bc[0, 0] in [2, 3]: # Neumann or Robbins
            A[ind, ind] = -(3 / (2 * dx) + bc[0, 2])
            A[ind, ind + 1] = 2 / dx
            A[ind, ind + 2] = -1 / (2 * dx)
            c[ind] = bc[0, 1]

    # 2. Upper x boundary (i=nx)
    for j in range(1, ny):
        ind = j * (nx + 1) + nx
        if bc[1, 0] == 1:
            A[ind, ind] = 1.0
            c[ind] = bc[1, 1]
        elif bc[1, 0] in [2, 3]:
            A[ind, ind] = (3 / (2 * dx) - bc[1, 2])
            A[ind, ind - 1] = -2 / dx
            A[ind, ind - 2] = 1 / (2 * dx)
            c[ind] = bc[1, 1]

    # 3. Lower y boundary (j=0)
    for i in range(1, nx):
        ind = 0 * (nx + 1) + i
        if bc[2, 0] == 1:
            A[ind, ind] = 1.0
            c[ind] = bc[2, 1]
        elif bc[2, 0] in [2, 3]:
            A[ind, ind] = -(3 / (2 * dy) + bc[2, 2])
            A[ind, ind + (nx + 1)] = 2 / dy
            A[ind, ind + 2 * (nx + 1)] = -1 / (2 * dy)
            c[ind] = bc[2, 1]

    # 4. Upper y boundary (j=ny)
    for i in range(1, nx):
        ind = ny * (nx + 1) + i
        if bc[3, 0] == 1:
            A[ind, ind] = 1.0
            c[ind] = bc[3, 1]
        elif bc[3, 0] in [2, 3]:
            A[ind, ind] = (3 / (2 * dy) - bc[3, 2])
            A[ind, ind - (nx + 1)] = -2 / dy
            A[ind, ind - 2 * (nx + 1)] = 1 / (2 * dy)
            c[ind] = bc[3, 1]

    # Corner nodes
    # A(1,1) -> A[0,0]
    A[0, 0] = 1; A[0, 1] = -0.5; A[0, nx + 1] = -0.5; c[0] = 0
    # A(nx+1, nx+1) -> A[nx, nx]
    A[nx, nx] = 1; A[nx, nx - 1] = -0.5; A[nx, 2 * (nx + 1) - 1] = -0.5; c[nx] = 0
    # Lower-left corner in indexing (ny*(nx+1)+1) -> A[ny*(nx+1), ny*(nx+1)]
    idx_ll = ny * (nx + 1)
    A[idx_ll, idx_ll] = 1; A[idx_ll, idx_ll + 1] = -0.5; A[idx_ll, idx_ll - (nx + 1)] = -0.5; c[idx_ll] = 0
    # Last node A(n,n) -> A[n-1, n-1]
    A[n-1, n-1] = 1; A[n-1, n-2] = -0.5; A[n-1, n - 1 - (nx + 1)] = -0.5; c[n-1] = 0

    # Solve the linear system
    u = np.linalg.solve(A, c)
    
    # Rearrange results in matrix form U(ny+1, nx+1)
    U = u.reshape((ny + 1, nx + 1))
    
    return x, y, U